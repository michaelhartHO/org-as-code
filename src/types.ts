// EventType is the category of data-point that can be stored.
export enum EventType {
  Skill = "skill",
  Person = "person",
  Team = "team",
}

export type EventTag = string;

export type EventsMap = Map<EventTag, RegistryData>;

export interface EventBase {
  tag: EventTag;
}

// The Skill, Person, and Team interfaces represent the data that Users will provide as data-points
export interface Skill extends EventBase {
  description: string;
  reference: string;
}
export interface Person extends EventBase {
  name: string;
  skills: [string, number][];
}
export interface Team extends EventBase {
  name: string;
  lead: string;
  services: string[];
}
export interface RegistryBase extends EventBase {
  type: EventType;
  date?: string;
}

// The SkillEvent, PersonEvent, and TeamEvent interfaces represent the data structures (built from data-points) that will be stored in the registry
export interface SkillEvent extends RegistryBase, Skill {
  type: EventType.Skill;
}

export interface PersonEvent extends RegistryBase, Person {
  type: EventType.Person;
}

export interface TeamEvent extends RegistryBase, Team {
  type: EventType.Team;
}

export type RegistryData = SkillEvent | PersonEvent | TeamEvent;
