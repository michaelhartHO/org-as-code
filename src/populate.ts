// populate.ts populates the database with the data-points

import {
  DbInsertRegistryDataFn,
  TimeSeriesDb,
} from "./timeSeriesDb.ts";
import { registrarFactory } from "./registrar.ts";
import { registerDataPoints } from "./registerDataPoints.ts";

export const db = new TimeSeriesDb();

export async function populateDatabaseFromPaths(
  dbInsertFn: DbInsertRegistryDataFn,
  paths: string[],
) {
  for (const path of paths) {
    await registerDataPoints(registrarFactory(dbInsertFn), path);
  }
}
