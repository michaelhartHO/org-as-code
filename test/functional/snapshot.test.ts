// Example of snapshot testing
// Update snapshots with:
//   $ deno test test/functional/snapshot.test.ts --allow-all -- --update
import { assertSnapshot } from "jsr:@std/testing/snapshot";

import { populateDatabaseFromPaths } from "../../src/populate.ts";
import {
  createDbInsertRegistryDataFn,
  TimeSeriesDb,
} from "../../src/timeSeriesDb.ts";
import { EventType } from "../../src/types.ts";

Deno.test("populateDatabaseFromPaths ok", async (t): Promise<void> => {
  const db = new TimeSeriesDb();
  const dbInsertFn = createDbInsertRegistryDataFn(db);

  const relativePath = "./test/functional/data-points";
  let dataPointsPath: string;
  try {
    dataPointsPath = Deno.realPathSync(relativePath);
  } catch (_error) {
    throw new Error(`Path does not exist: ${relativePath}`);
  }

  const paths = [
    `${dataPointsPath}/skills`,
    `${dataPointsPath}/teams`,
    `${dataPointsPath}/people`,
  ];
  await populateDatabaseFromPaths(dbInsertFn, paths);

  const skillsDb = db.getEventsDb(EventType.Skill);
  await assertSnapshot(t, skillsDb);

  const teamsDb = db.getEventsDb(EventType.Team);
  await assertSnapshot(t, teamsDb);

  const peopleDb = db.getEventsDb(EventType.Person);
  await assertSnapshot(t, peopleDb);
});
