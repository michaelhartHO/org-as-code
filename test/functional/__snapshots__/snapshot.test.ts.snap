export const snapshot = {};

snapshot[`populateDatabaseFromPaths ok 1`] = `
Map(2) {
  "Pachinko" => {
    date: "1999-02-12",
    description: "Pachinko master",
    reference: "https://en.wikipedia.org/wiki/pachinko?variant=zh-cn",
    tag: "Pachinko",
    type: "skill",
  },
  "Pokémon" => {
    date: "2001-12-31",
    description: "Pokémon master",
    reference: "https://www.pokemon.com",
    tag: "Pokémon",
    type: "skill",
  },
}
`;

snapshot[`populateDatabaseFromPaths ok 2`] = `
Map(1) {
  "Team Rocket" => {
    date: "2021-01-01",
    lead: "keyser.söze",
    name: "Launch Capabilities",
    services: [
      "Engines",
      "Payload",
    ],
    tag: "Team Rocket",
    type: "team",
  },
}
`;

snapshot[`populateDatabaseFromPaths ok 3`] = `
Map(1) {
  "keyser.söze" => {
    date: "1995-08-28",
    name: "Keyser Söze",
    skills: [
      [
        "Pachinko",
        7,
      ],
      [
        "Pokémon",
        9,
      ],
    ],
    tag: "keyser.söze",
    type: "person",
  },
}
`;
